Add topic-list page

PIG208 opened the pull request:
Stacked on top of #1491.

Some non-goals of this change are deferred to #1499. In this implementation, we fetch the topics but do not handle all events to receive live-updates. It's expected that when topics are resolved/unresolved or moved, or when new messages arrived, the changes to the topic-list will not be seen until the next fetch.

We also skip thinning down the app bar, since that will require work on app bars on message-list page too.

The PR is structured to encourage side-by-side comparison with similar existing code. Namely `_TopicItem` from `lib/widgets/inbox.dart` and `MessageListAppBarTitle`.

Fixes: #1158

<details>
<summary>
screenshots (taken on my Android device, hence the left-aligned app bar!)
</summary>

| | light | dark |
| - | - | - |
| regular | ![image](https://github.com/user-attachments/assets/166717f8-6a13-486f-9a79-c3247f0b7b5a) | ![image](https://github.com/user-attachments/assets/7e253fd4-895a-4ef3-ad58-7d156142c164) |
| unsubscribed | ![image](https://github.com/user-attachments/assets/b6ef8226-b791-4d60-a4cb-019fa3c8f2c0) | ![image](https://github.com/user-attachments/assets/159cc573-8123-4dce-8d1d-7642075129bb) |
| unknown channel | ![image](https://github.com/user-attachments/assets/58b152cf-ad13-48dd-982e-81cc1de8be4b) | ![image](https://github.com/user-attachments/assets/00cf0ec8-cc8a-4bc0-a245-6f38a7e49b93) |

| small | regular | large |
| - | - | - |
| ![image](https://github.com/user-attachments/assets/6927c894-92ee-4126-b78c-6fc8918f7de3) | ![image](https://github.com/user-attachments/assets/1ec9a1ec-963c-4e15-b35d-95f23e994cca) | ![image](https://github.com/user-attachments/assets/f11397ba-68ce-47b9-8a57-656f3939ac36) |

| message-list | channel action sheet |
| - | - |
| ![image](https://github.com/user-attachments/assets/d1e5781c-34cf-41ae-95c1-9dd73d6dafd8) | ![image](https://github.com/user-attachments/assets/a37ce2d4-77e5-4a33-b21c-229ff4602ec5) |

| non-muted topic | muted topic |
| - | - |
| ![image](https://github.com/user-attachments/assets/166717f8-6a13-486f-9a79-c3247f0b7b5a) | ![image](https://github.com/user-attachments/assets/35ab5267-35a1-440d-89cf-c89bd6e1b839) |

`debugDefaultTargetPlatformOverride = TargetPlatform.iOS`:

<img src="https://github.com/user-attachments/assets/94079d44-6418-4177-b6b9-0363043da53e" width=200 />


</details>

PIG208 commented:
Updated the PR to implement a intentional deviation from the Figma design on aligning items, discussed in [#mobile-team > topic list item alignment @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/topic.20list.20item.20alignment/near/2173252).

alya commented:
For the "small" and "large" screenshots, I think it would look much better if the checkmarks scaled too.

alya commented:
Let's name the option "List of topics". I think that's a bit easier to parse, and it's what we're currently testing for the web app UI.

PIG208 commented:
Thanks both! Updated the PR and the screenshots. I picked 1.5 to be the `maxScaleFactor` for icons, since that appears to be what we are using in most other places.

I clamped only the icons, though, not the topic name text. Because I think #1023 applies to  scaling relatively unimportant information.

alya commented:
Works for me, thanks!

chrisbobbe commented:
The app bar's back button and "Channel feed" button are broken; sometimes when I tap it doesn't respond and this gets logged:

<details>

```
======== Exception caught by foundation library ====================================================
The following assertion was thrown while dispatching notifications for WidgetStatesController:
setState() or markNeedsBuild() called when widget tree was locked.

This _IconButtonM3 widget cannot be marked as needing to build because the framework is locked.
The widget on which setState() or markNeedsBuild() was called was: _IconButtonM3
  style: ButtonStyle#2426d(mouseCursor: WidgetStateMapper<MouseCursor?>({WidgetState.disabled: null, WidgetState.any: null}))
  dependencies: [IconButtonTheme, IconTheme, InheritedCupertinoTheme, _InheritedTheme, _LocalizationsScope-[GlobalKey#2fb94]]
  state: _ButtonStyleState#9e3dc
When the exception was thrown, this was the stack: 
#0      Element.markNeedsBuild.<anonymous closure> (package:flutter/src/widgets/framework.dart:5273:9)
#1      Element.markNeedsBuild (package:flutter/src/widgets/framework.dart:5283:6)
#2      State.setState (package:flutter/src/widgets/framework.dart:1219:15)
#3      _ButtonStyleState.handleStatesControllerChange (package:flutter/src/material/button_style_button.dart:326:5)
#4      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:435:24)
#5      WidgetStatesController.update (package:flutter/src/widgets/widget_state.dart:1152:7)
#6      _InkResponseState.updateHighlight (package:flutter/src/material/ink_well.dart:995:26)
#7      _InkResponseState.handleTapCancel (package:flutter/src/material/ink_well.dart:1211:5)
#8      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:345:24)
#9      TapGestureRecognizer.handleTapCancel (package:flutter/src/gestures/tap.dart:800:11)
#10     BaseTapGestureRecognizer._checkCancel (package:flutter/src/gestures/tap.dart:388:5)
#11     BaseTapGestureRecognizer.resolve (package:flutter/src/gestures/tap.dart:336:7)
#12     OneSequenceGestureRecognizer.dispose (package:flutter/src/gestures/recognizer.dart:470:5)
#13     PrimaryPointerGestureRecognizer.dispose (package:flutter/src/gestures/recognizer.dart:765:11)
#14     RawGestureDetectorState.dispose (package:flutter/src/widgets/gesture_detector.dart:1529:18)
#15     StatefulElement.unmount (package:flutter/src/widgets/framework.dart:5922:11)
#16     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2075:13)
#17     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#18     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#19     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#20     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#21     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:6994:14)
#22     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#23     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#24     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#25     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#26     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#27     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#28     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#29     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#30     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:6994:14)
#31     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#32     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#33     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:6994:14)
#34     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#35     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#36     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#37     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#38     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#39     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#40     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#41     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#42     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#43     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#44     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#45     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#46     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#47     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#48     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#49     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#50     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#51     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#52     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#53     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#54     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#55     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#56     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#57     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#58     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#59     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#60     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#61     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#62     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#63     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#64     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#65     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#66     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#67     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#68     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#69     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#70     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#71     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#72     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#73     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#74     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#75     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#76     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#77     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#78     ComponentElement.visitChildren (package:flutter/src/widgets/framework.dart:5763:14)
#79     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#80     _InactiveElements._unmount.<anonymous closure> (package:flutter/src/widgets/framework.dart:2073:7)
#81     SingleChildRenderObjectElement.visitChildren (package:flutter/src/widgets/framework.dart:6994:14)
#82     _InactiveElements._unmount (package:flutter/src/widgets/framework.dart:2071:13)
#83     ListIterable.forEach (dart:_internal/iterable.dart:49:13)
#84     _InactiveElements._unmountAll (package:flutter/src/widgets/framework.dart:2084:25)
#85     BuildOwner.lockState (package:flutter/src/widgets/framework.dart:2965:15)
#86     BuildOwner.finalizeTree (package:flutter/src/widgets/framework.dart:3288:7)
#87     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:1266:19)
#88     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:495:5)
#89     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1438:15)
#90     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1351:9)
#91     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1204:5)
#92     _invoke (dart:ui/hooks.dart:331:13)
#93     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:444:5)
#94     _drawFrame (dart:ui/hooks.dart:303:31)
The WidgetStatesController sending notification was: WidgetStatesController#d045d({})
====================================================================================================
```

</details>

Looks like an upstream regression, https://github.com/flutter/flutter/issues/168445.

chrisbobbe commented:
We should use less than 28px padding at the start of each topic row. From the issue description:

> For this issue, differences from the design there:
> 
> - The design shows a "pinned topics" feature, which is a possible future Zulip feature but doesn't currently exist. Pending that feature, we'll leave out the "pinned" icon, and avoid using up horizontal space for it.

chrisbobbe commented:
For the layout, if we use an empty box as a placeholder for an icon when it's absent, it should scale exactly the same way as the icon does. See the start of the topic text being misaligned in the "large" screenshot:

<img width="79" alt="image" src="https://github.com/user-attachments/assets/50cca560-c41b-4947-9104-5181aaa5b8e7" />

chrisbobbe commented:
I started a discussion about the chevron-down icon in the app bar: [#mobile-team > topic list: chevron-down icon in app bar? @ 💬](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/topic.20list.3A.20chevron-down.20icon.20in.20app.20bar.3F/near/2175960)

PIG208 commented:
Thanks for the review! I adjusted the padding at the start of each row to 6px (matching the start padding the "pinned" icon would have), and addressed the other issues. The screenshots have been updated.

> Looks like an upstream regression, https://github.com/flutter/flutter/issues/168445, that's being worked on.

Thanks for looking into this! Yeah, for this one, we might need https://github.com/flutter/flutter/pull/168546 to land first.

gnprice commented:
> The app bar's back button and "Channel feed" button are broken; sometimes when I tap it doesn't respond

AFAICT the symptoms of that issue don't reproduce in a release build; the buttons work fine the first time. (I spent a few minutes just now trying to reproduce, in order to evaluate whether the issue will affect the [upcoming beta v0.0.29](https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/beta.20release.20management/near/2174231).)

In a debug build, I also see it on other screens even without this PR.

So definitely an annoyance in dev — plus might be a sign of further bugginess that's just harder to spot the symptoms of — and will be good to see fixed upstream. But I don't think it needs to interact with merging this PR.

chrisbobbe commented:
When I rebase this onto current `main`, the second commit has some `tools/check` errors:

<details>

```
$ tools/check analyze l10n
Running analyze...
Analyzing zulip-flutter...                                              

  error • Missing concrete implementation of 'getter abstract class
         ZulipLocalizations.topicsButtonLabel' •
         lib/generated/l10n/zulip_localizations_de.dart:8:7 •
         non_abstract_class_inherits_abstract_member

1 issue found. (ran in 2.8s)
Running l10n...
Error: there were updates to l10n:
 M lib/generated/l10n/zulip_localizations_de.dart

FAILED: analyze l10n

To rerun the suites that failed, run:
  $ tools/check analyze l10n
```

</details>

Could you rebase and fix those please?

PIG208 commented:
Thanks! Just updated this. I found that it also happens to the last commit. All commits that contain changes adding new strings to translate are probably affected, since we just recently added `de` (#1522).

chrisbobbe commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,333 @@
+    for (final GetStreamTopicsEntry(:maxId, name: topic) in lastFetchedTopics!) {
+      final unreadMessageIds =
+        unreadsModel!.streams[widget.streamId]?[topic] ?? <int>[];
+      final countInTopic = unreadMessageIds.length;
```
Let's centralize on `Unreads.countInTopicNarrow` for this; I could imagine wanting to refine it for something like muted senders etc.

PIG208 replied:
We might still need to keep `unreadMessageIds` to compute the value of `hasMention`.
Looks like inbox page does something like that with DMs:

```dart
      final countInNarrow = unreadsModel!.countInDmNarrow(dmNarrow);
      if (countInNarrow == 0) {
        continue;
      }
      final hasMention = unreadsModel!.dms[dmNarrow]!.any(
        (messageId) => unreadsModel!.mentions.contains(messageId));
```

chrisbobbe commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,333 @@
+              // A null [Icon.icon] makes a blank space.
+              _IconMarker(icon: topic.isResolved ? ZulipIcons.check : null),
+              Expanded(child: Opacity(
+                opacity: isTopicVisibleInStream ? 1 : 0.5,
```
This looks a little odd—the value (from the relevant dartdoc) is about "whether" this topic should appear, not how it should appear. Is this logic correct, and is that dartdoc on `store.isTopicVisibleInStream` correct?

PIG208 replied:
The topics with "isTopicVisibleInStream=false" will be hidden on web, in the list that appears when you just click on the channel.

This topic-list however should probably match the topic-list from "show all topics" button on web. Because in the Figma design, muted topics in the topic-list appear transparent. I think similar to web, this offers a way for the user to access the muted topics.

In conclusion, both are correct, but we should probably name the bool differently and comment on why we use the helper this way.

gnprice replied:
Yeah, that value means whether it should appear in the channel feed, the message list. That'd be good to clarify in its doc.

Here in the list of topics for the channel, all topics are shown.

PIG208 replied:
Or maybe it is more appropriate to use `topicVisibilityPolicy` directly, and compare the value to `UserTopicVisibilityPolicy.muted`.

gnprice replied:
Yeah, that's probably a good way to write this code.

chrisbobbe commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,333 @@
+                spacing: 4,
+                children: [
+                  ...trailingWidgets,
+                  if (trailingWidgets.isEmpty)
+                    const SizedBox(width: 53),
```
I see this is what the Figma is doing, but it feels odd: it doesn't create a consistent end margin for the topic text, because the gap will be less than 53px when there's just one or two `trailingWidgets`, right? How about just removing this and allowing a longer line for the topic text.

chrisbobbe commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,333 @@
+  }
+}
+
+// This is adapted from [UnreadCountBadge].
+class _UnreadCountBadge extends StatelessWidget {
+  const _UnreadCountBadge({required this.count});
```
It really seems like we should be able to use `UnreadCountBadge` directly here—if the Figma has inconsistencies among unread-count badges, I'd suspect they're accidental; could you investigate?

PIG208 replied:
Yeah, it would be ideal to share it.

The `UnreadCountBadge` we know today comes from Nov 2023 (d9eb0d388), when it was extracted from `RecentDmConversationsPage`. For DM, the [current](https://www.figma.com/design/1JTNtYo9memgW7vV6d0ygq/Zulip-Mobile?node-id=4918-2143&m=dev) Figma design's unread count badge looks different.

I assume that the design for inbox page and subscription-list page got outdated this way too. Since we haven't got to implementing those redesigns yet, just updating `UnreadCountBadge` to use the new design might make it out-of-place in those pages.

Relatedly, if we can, I would like to extract `_TopicItem` from `lib/widgets/inbox.dart` for reuse too. When I tried that while drafting this, it turned out that we would need to implement more of the inbox-page redesign to make the shared widget fit.

I think a reasonable strategy will be implementing these widgets in topic-list first, while referring to the code structure of the existing ones (to compare what's changed). We can consider extracting these widgets once we get to implementing other redesigned  pages, and phase out `UnreadCountBadge` in this process.

chrisbobbe replied:
Sure, that sounds reasonable. Could you leave a TODO for this plan?

PIG208 replied:
Yeah, I plan to file some issues later.

PIG208 replied:
Opened #1527, and I found that #1406 already exists, nice.

chrisbobbe commented on test/widgets/topic_list_tests.dart:
```diff
@@ -0,0 +1,327 @@
+    await tester.pumpWidget(TestZulipApp(
+      accountId: eg.selfAccount.id,
+      child: TopicListPage(streamId: channel.streamId)));
+    await tester.pumpAndSettle();
```
Can this be done with individual `tester.pump`s rather than a `pumpAndSettle`?

chrisbobbe commented on test/widgets/topic_list_tests.dart:
```diff
@@ -0,0 +1,327 @@
+        eg.getStreamTopicsEntry(name: unresolvedTopic.apiName),
+      ]);
+
+      assert(resolvedTopic.displayName == '✔ resolved', resolvedTopic.displayName);
+      check(findInTopicItemAt(0, find.text('✔ resovled'))).findsNothing();
```
oops: should be `find.text('✔ resolved')`

chrisbobbe reviewed (commented):
Thanks! Small comments below, and it looks like there's a (nearly?) resolved discussion about making this page more accessible from a message-list page for a topic narrow: [#mobile > Flutter beta: missing topic list @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Flutter.20beta.3A.20missing.20topic.20list/near/2177515)

PIG208 commented:
Thanks! Updated the PR.

chrisbobbe commented:
Thanks, looks good! Marking for Greg's review.

gnprice reviewed (commented):
Thanks to you both! Generally this looks good; comments below.

I'm about to go AFK, but I've gotten up through the non-test changes in the second commit:
39b672418 topics: Add topic list page

Still ahead for me to read are the tests, and the third commit:
33fc01d99 topics: Add TopicListButton to channel action sheet

chrisbobbe commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,326 @@
+    // This is adapted from parts of the build method on [_InboxPageState].
+    final topicItems = <_TopicItemData>[];
+    for (final GetStreamTopicsEntry(:maxId, name: topic) in lastFetchedTopics!) {
+      final countInTopic =
+        unreadsModel!.countInTopicNarrow(widget.streamId, topic);
+      final unreadMessageIds =
+        unreadsModel!.streams[widget.streamId]?[topic] ?? <int>[];
+      final hasMention = unreadMessageIds.any((messageId) =>
```
Oh! it was my suggestion to avoid `unreadMessageIds.length`—https://github.com/zulip/zulip-flutter/pull/1500#discussion_r2101327987

I'm okay to be overruled on that though :)

PIG208 commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,326 @@
+            //   https://chat.zulip.org/#narrow/channel/243-mobile-team/topic/topic.20list.20item.20alignment/near/2173252
+            children: [
+              // A null [Icon.icon] makes a blank space.
+              _IconMarker(icon: topic.isResolved ? ZulipIcons.check : null),
```
No, because it's not extra-faded in the Figma design, (this is also why we couldn't wrap the entire row in a single `Opacity` widget).

PIG208 commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,326 @@
+                    fontStyle: topic.displayName == null ? FontStyle.italic : null,
+                    color: designVariables.textMessage,
+                  ),
+                  topic.unresolve().displayName ?? store.realmEmptyTopicDisplayName))),
```
Hm, I thought the example text  "[…] which could be more than  2 lines if we want" meant that we can have >2 lines with no upperbound. But I guess without baseline alignment, more than 3 lines can look odd. Let's go with `maxLines: 3` now and revisit this if we fix the alginment.

PIG208 replied:
Opened #1528 for implementing baseline alignment.

PIG208 commented on lib/widgets/topic_list.dart:
```diff
@@ -0,0 +1,326 @@
+      return bMaxId.compareTo(aMaxId);
+    });
+
+    return ListView.builder(
```
Yeah, and we also want to skip padding `bottom` here, because `ListView` handles it.

PIG208 commented on lib/widgets/message_list.dart:
```diff
@@ -284,6 +287,35 @@ class _MessageListPageState extends State<MessageListPage> implements MessageLis
+      },
+      behavior: HitTestBehavior.opaque,
+      child: Padding(
+        padding: EdgeInsetsDirectional.fromSTEB(4, 8, 12, 8),
```
Went with 12px, and posted a screenshot in [#mobile > Flutter beta: missing topic list @ 💬](https://chat.zulip.org/#narrow/channel/48-mobile/topic/Flutter.20beta.3A.20missing.20topic.20list/near/2179752):

<img src="https://github.com/user-attachments/assets/257bf374-0869-4834-b164-e483daa52d77" width=200 />

gnprice commented on test/widgets/action_sheet_test.dart:
```diff
@@ -227,6 +227,7 @@ void main() {
       void checkButtons() {
         check(actionSheetFinder).findsOne();
         checkButton('Mark channel as read');
+        checkButton('List of topics');
```
nit: match the order these appear in the action sheet (and in the code, and the tests below)
```suggestion
        checkButton('List of topics');
        checkButton('Mark channel as read');
```

gnprice commented on test/widgets/topic_list_tests.dart:
```diff
@@ -0,0 +1,332 @@
+import '../stdlib_checks.dart';
+import 'test_app.dart';
+
+void main() {
```
Filename should end in `_test.dart` so that `flutter test` picks it up :slightly_smiling_face: 
```
+++ test/widgets/topic_list_tests.dart
```

gnprice commented on test/widgets/topic_list_tests.dart:
```diff
@@ -0,0 +1,332 @@
+      final resolvedTopic = TopicName('resolved').resolve();
+      final unresolvedTopic = TopicName('unresolved');
+      await prepare(tester, topics: [
+        eg.getStreamTopicsEntry(name: resolvedTopic.apiName),
+        eg.getStreamTopicsEntry(name: unresolvedTopic.apiName),
+      ]);
+
+      assert(resolvedTopic.displayName == '✔ resolved', resolvedTopic.displayName);
+      check(findInTopicItemAt(0, find.text('✔ resolved'))).findsNothing();
```
Several of these test cases are a bit fragile because they let `maxId` be the same for several items in the list (which is also unrealistic — the message with that ID can belong to at most one conversation), and then their checks expect a particular order for the items to appear in.

Can fix by just giving explicit `maxId: 2`, `maxId: 1`, etc.

gnprice commented on test/widgets/topic_list_tests.dart:
```diff
@@ -0,0 +1,332 @@
+    check(find.text('topic B')).findsOne();
+  });
+
+  group('_TopicItem', () {
```
nit: this group isn't really specific to _TopicItem — a lot of the logic it's testing is in `_TopicListState`. (In fact I saw this group and that there weren't any other tests left in the file, and then thought that meant there wouldn't be tests for what's in `_TopicListState.build` and wondered if I should ask for some.)

I think the group can just be unwrapped, and its contents de-indented to the top level of `main`. One accurate name for the group would be "page body"… but the whole file is about this particular page, so it seems reasonable for the page body to take the top level.

gnprice reviewed (commented):
Thanks for the revision! Those changes all look good.

I've just read through the rest of the branch, and generally it all looks good; a few comments below.
